### PR TITLE
Add 'coords_ref' option to 'add_points' to specify lon/lat or pixel coordinates

### LINKED ISSUE #55: Additional text overlay options
Suggested improvements to text overlays:

* Overlays can be specified using `points` and/or `text` (and you can easily add other names if you want more), so that you can specify different rendering attributes for each set of points/text
* The symbol can be `None` (you need to set `ptsize=0` if you don't want a gap) so you can have a pure text overlay
* Coordinates can be specified in 'pixel' space or 'lonlat' space (the default is `lonlat`) by giving an additional key in the tuple `((100,50), "my text", "pixel")`  Negative coords go from right/bottom.

### PR DESCRIPTION
**Pull request by howff**
Suggested improvements to text overlays:

* Overlays can be specified using `points` and/or `text` (and you can easily add other names if you want more), so that you can specify different rendering attributes for each set of points or text
* The symbol can be `None` (you need to set `ptsize=0` if you don't want a gap) so you can have a pure text overlay
* Coordinates can be specified in 'pixel' space or 'lonlat' space (the default is `lonlat`) by giving an additional key in the tuple `((100,50), "my text", "pixel")`  Negative coords go from right/bottom. Backwards compatible.

 - [X] Closes #55
 - [ ] Tests added
 - [X] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [X] Fully documented in the source code

### CODE REVIEW

**Comment by howff** [comment]
You are right to question the use of the word "text", we could just as easily have "points", "points1", "points2", "points3", ... "pointsN".
Alternatively we could make the default value of "symbol" and "ptsize" zero for "text".

**Review by adybbroe — COMMENTED** [review]
LGTM, thanks for the work.
For the doc strings you could use longer line lengths (and then less lines). Think we said that for coding style we should stay within 120 characters (rather than just 80) for maximum line length.

**Comment by adybbroe** [comment]
It could be nice with a small use case example here, or preferably at the doc pages. Again, nice feature addition, and thanks for the work!

File: pycoast/cw_base.py
Comment (yufeizhu600) [inline pycoast/cw_base.py:1069]: Since the `(lon, lat)` in the doc string refers to `(longitude, latitude)` or `(x, y)` in this change; I was wondering should it be changed to some thing else like `(coord1, coord2)` to avoid misunderstanding.

File: pycoast/cw_base.py
Comment (yufeizhu600) [inline pycoast/cw_base.py:63]: May I suggest to change the name of `x`, `y` in the function signature? Since the meaning of their value have different meaning depend on the `coord_ref`, I think it is better to use some generic names instead of `x, y` or `lon, lat` here.

Comment (djhoese) [reply pycoast/cw_base.py:63]: I agree with the change in the docstring, but maybe `x` and `y` are generic enough while still referring to the dimension/axis being used.

File: pycoast/cw_base.py
Comment (howff) [reply pycoast/cw_base.py]: well spotted :-) thank you, fix now pushed

File: pycoast/cw_base.py
Comment (howff) [reply pycoast/cw_base.py:63]: I think x,y are the generic axis names.
I've changed the docstring and code to use x,y everywhere appropriate now.

**Comment by yufeizhu600** [comment]
Thanks for updating @howff, LGTM. 
Don't forget to add your name into the AUTHORS.md file.

**Review by yufeizhu600 — APPROVED** [review]

**Comment by djhoese** [comment]
I'm really sorry about not merging this. I'm not sure what happened.

I'm sorry to ask this so late after so many approvals, but would it be possible to add some level of testing for this logic?

**Comment by howff** [comment]
Sorry I'm having trouble with the tests. About 7 out of 33 fail (on the main branch, not this PR branch). The test images look visually identical to me but their rms larger than max_value so they fail. Computing a magick compare difference image makes it look like all the coastline points are just slightly different for some reason.
```
FAILED test_pycoast.py::TestPIL::test_add_points_pil - AssertionError: False is not true : Writing of nh points failed
FAILED test_pycoast.py::TestPIL::test_config_file_points_and_borders_pil - AssertionError: False is not true : Writing of nh poi...
FAILED test_pycoast.py::TestPILAGG::test_add_points_agg - AssertionError: False is not true : Writing of nh points failed
FAILED test_pycoast.py::TestPILAGG::test_config_file_coasts_and_grid - OSError: cannot load font
FAILED test_pycoast.py::TestPILAGG::test_config_file_points_and_borders_agg - OSError: cannot load font
FAILED test_pycoast.py::TestPILAGG::test_grid_agg_txt - AssertionError: False is not true : Writing of grid failed for AGG
FAILED test_pycoast.py::TestPILAGG::test_grid_nh_agg - AssertionError: False is not true : Writing of nh grid failed for AGG
```
For this reason I've not tried the tests on this PR branch yet.

**Comment by djhoese** [comment]
The tests for this package are notoriously picky/rigid. Would it be possible to add a simple test that maybe doesn't compare the output to an expected image, but instead just runs through the various code branches to make sure they execute as expected? If it makes sense you could either use `mock` to wrap some of the lower-level functions to make sure they get called with the right parameters (not my preference) or generate one image with one set of arguments and compare it to the image generated with the same arguments but in different units. For example, add some text or points with lon/lat degrees coordinates and save the image, then do the same but with `coord_ref` set to `image` and the equivalent row/column coordinates. The images should be the same so that is easy to confirm. You could even change the coord_ref but not the point coordinates and make sure the image changed.

**Comment by lobsiger** [comment]
@howff  may I suggest to reconsider your own proposal to use 'points', 'points1', 'points2', 'points3' as param_keys? After all it's calling add_points(). I now use the same logic for 'cities' that is calling add_cities() and I already implemented as a quick hack 'shapefile' calling add_shapefile() for Christian Peters (he draws e.g. German state boundaries with a shape file). Using more than on shape file in the dict could again be implemented as 'shapefile', ''shapefile1', 'shapefile2', 'shapefile3'. And then there is still PyCoast add_line() and add_polygon() that could be implemented to be called the same way from satpy with an overlay dict.

**Comment by djhoese** [comment]
I may be forgetting how all of this works, but if we're talking about a series of "things" then the logic should accept a list instead of the added complexity of adding a numeric suffix to the key.

**Comment by lobsiger** [comment]
Yes, I was playing with lists which means the user can have as many kinds of 'points' and 'cities' he likes. But the problem of this approach is that it will not work with configuration *.ini files. And I was also thinking about adding this PyCoast feature for Satpy.
And TBH, I cannot think of a common PyTROLL/SatPy use case where four kinds of 'points' and four kinds of 'cities' will not do ...

**Comment by howff** [comment]
I don't quite understand what you're getting at, Ernst. satpy currently allows a list of text strings to be plotted at lat,lon locations, so this PR simply allows an additional list of text strings to be plotted at locations specified by pixel x,y coordinates. It's simply a way of adding an extra coordinate system.

**Comment by lobsiger** [comment]
@howff I'm talking about your alias 'points' and 'text'. This allows to add to kinds of points each with its parameter set, symbol and points_list. So I'm talking about a list of such parameter sets that I cannot see how to get working with *.ini files. Instead of having two sets 'points' and 'text' I propose four sets 'points', 'points1', 'points2', 'points3' like your own comment 2 Nov 2021 above. See also my image in PR #61 that uses 'cities' and 'cities1'. Adding cities with pixel coordinates makes no sense though.

**Comment by djhoese** [comment]
> But the problem of this approach is that it will not work with configuration *.ini files

The `pointsN` approach is just not flexible. I'd rather support the `pointsN` sections in `.ini` only and pass those as a list of `points` to the methods that use them. In the future I could see pycoast depending on YAML configuration files like other pytroll tools do.

> And TBH, I cannot think of a common PyTROLL/SatPy use case where four kinds of 'points' and four kinds of 'cities' will not do ...

The second you say something like this or implement it is when someone comes along and says "I need 5 kinds". Aren't styles controlled with this? What about wanting a color for each category of a cloud product and show a city's "cloud type" with that color.

**Comment by lobsiger** [comment]
@djhoese this PyCoast PR number 56 "Additional text overlay options" among other things has the magic lines:

  for param_key in ['points', 'text']:
            if param_key not in overlays:
                continue
  ...
  params = overlays[param_key].copy()
  ...
That lets you use two styles of points in overlays when saving your scene in SatPy (without the need of using PyDecorate). It seamlessly integrates in PyCoast function add_overlay_from_dict() and even works with the rather limited *.ini parser used by add_overlay_from_config(). This PR has been reviewed/approved by three developers including yourself writing on 2 Nov 2021:
<cite>
This looks good to me. I'm curious if we need the text alias for points, but I guess it makes sense that you might want to overlay a series of points and a series of text points so having that is nice. Just thinking "out loud".
</cite>

While you even questioned the need for 2 aliases you now say 4 aliases might not be enough and we should rewrite the config parser and about everything. Inspector Columbo would certainly say: "I'm confused ..."

**Comment by djhoese** [comment]
I know what I said. I would like to keep this PR as it is for now. The points and text alias is fine for now as it at least gives you two types of categories of styles. Going more than that and talking about multiple styles per type of point is a different use case. That work though should be done *after* this PR and *after* all your other PRs.

I am completely swamped right now and do not have time to look at all of these features and changes right now. I will get to these things as quickly as I can. What I need for this PR is a couple tests as I described here:

https://github.com/pytroll/pycoast/pull/56#issuecomment-1084771553

If we can get those and the tests pass on CI then we can merge this and move forward on the other PRs related to the changes here. If @howff does not have time to add the tests and you @lobsiger have a desire to see this merged so you can continue your own work then feel free to comment here with a test(s) and I can add it to this branch.

**Comment by djhoese** [comment]
I just merged current main into this and will try adding some of the tests I suggested.

**Comment by djhoese** [comment]
@mraspaud @howff @lobsiger Ok I think this PR is ready for re-review. Here's some of what I did:

1. Add a few add_points tests that just verify that passing coordinates with lon/lat coordinates get the same result when image col/row coordinates are provided. I also added some tests for when parameters or coordinates are wrong. This revealed at least one bug where you could request a point at (num_rows, num_cols) and it would work (off by one error).
2. I rearranged some things so that error handling was more as I expected. Mainly, `coord_ref` being an invalid value would just result in no points being added and a log message. I expect a hard exception in this case.
3. I have *a lot* of other changes I want to make, but they would interfere pretty heavily with @lobsiger's #61. I'll wait until we merge that one (or modify that branch directly) to make that work.
4. I added an initial pre-commit config that can be installed with `pre-commit install` to check flake and other issues at commit time. However I highly recommend that no one actually install this until I've had a chance to rewrite a lot of pycoast. You basically won't be able to make a commit without turning of pre-commit or rewriting pycoast otherwise.
5. I fixed coverage reporting so this repository's CI is now reporting coverage to coveralls properly.

**Comment by djhoese** [comment]
As part of the "I have a lot of other changes" thing, I'm not going to fix the CodeFactor problem until I can do all of the other changes.

**Comment by lobsiger** [comment]
We should keep in mind that PR 56, 60, 61 and 62 are all fixes/enhancements of add_overlay_from_dict() that is used when saving images with overlays from satpy. In the long run and from the satpy user perspective I would prefer that fonts are specified the same way (using the same parameters and naming) for grid, points and cities. The 'cities' PR was started by copying the 'points' code. The user should see minimal differences when preparing different styles of 'points' or 'cities'. Basically this boils down to points_list (plus the enhancement of for choosing pixel coordinates) and cities_list. These must remain the only differences the satpy user sees. That said 'points' should finally also use (@djhoese that was your own proposal) the more flexible list approach of PR 61 and PR 62 instead of the alias 'points'/'text' approach.

One thing I didn't touch but should be reviewed is the usage of colors and opacities. As it is now the font color is always identical with the 'outline' of the symbol. There is an 'outline_opacity' but this is not used by agg fonts. This means using 'points' or 'text' with symbol=None and ptsize=0 there is no way to generate an agg 'water mark'. The easiest way would be to use opacity=outline_opacity in den definition of agg fonts as well. Maybe a better way would be to have, fully independant from the rest of the used colors, 'font_file', 'font_fill', 'font_opacity' and 'font_size'.

I think PR 56 and PR 61 (as both pass the tests) can be merged as is and then the code factor problem should be handled by simplifying/refactoring commonly used code blocks. After that I see no need for big rewrites of PyCoast anymore. I have added more meaningful cities_from_dict tests and also changed one opaycity to make agg fill like pil.

**Comment by djhoese** [comment]
> In the long run and from the satpy user perspective I would prefer that fonts are specified the same way (using the same parameters and naming) for grid, points and cities. 

I agree and I don't think anyone would disagree with wanting consistency. It is great that you're worried about the behavior from Satpy's point of view, but we have to keep in mind that Satpy is not the only interface to this functionality. This is an open source library that has been available to the public for many years. No matter how bad and inconsistent the interfaces are or how buggy the code is, we still have to provide some period of transition between versions...or at least try. For example, I distribute software that includes a script that uses pycoast and has command line flags for a lot of these options without using Satpy at all. I'm a special case in that I can probably update the script pretty fast and my users won't know the difference, but I might not be the only user.

> Maybe a better way would be to have, fully independant from the rest of the used colors, 'font_file', 'font_fill', 'font_opacity' and 'font_size'.

I like this idea. This does mean, however, that there would be a lot of new options for some parts of pycoast. If we weren't worried about `.ini` support I might say we should have our own Font class in pycoast that wraps all of these options.

> can be merged as is and then the code factor problem should be handled by simplifying/refactoring commonly used code blocks. After that I see no need for big rewrites of PyCoast anymore.

I'm not saying you need to do any of this work. I'll do it. My point is that pycoast is ugly. It was written a long time ago before better coding practices were known by the authors. The Pytroll group has better coding styles, standards, and techniques that we try to stick to nowadays that should be applied here. This type of refactoring should reveal better solutions to some of the issues and inconsistencies you've identified and have been a problem for a long time. The code has an overall bad design and is difficult to maintain and is slow because of it.

Overall, I want to say that we (the people who maintain this library) understand that there is a lot that isn't working in this package for our users. But I also want it understood that we didn't do this on purpose and we don't like that it is the way it is. The bottom line is that pycoast mostly works for the use cases that people originally needed. A lot of the original authors of pycoast and pydecorate aren't involved in the project any more so this project has been left dormant as we slowly learn how it works and what is wrong with it. We do this in our free time because most of our employers don't need extra functionality or aren't encountering any of the bugs that exist. I say all of this because I'm not sure how to take some of your comments. It feels like you are telling me what to do and blaming me for the way this project is. I get it, but also understand that we're doing what we can with the time and energy that we have.

**Comment by lobsiger** [comment]
@djhoese I'm certainly blaming no one. English is a foreign language for me, so this might lead to misunderstandings then and when. I started using Python and satpy 2018 and have been happy with it from day one. Until recently, when I finally decided to open a github account, I had no idea of all these modern programming methods. I did these PyCoast PRs because I had issues with satpy scripts that I have wiritten for EUMETCast and that a couple of other EUMETCast users work with now. And I noted that PyCoast was dormant as the initial programmers went away. I just wanted to help and prefer small steps instead of major leap steps. Another reason I decided to learn some Git in my old days is because I wrote a couple of programs 40 years ago to calculate aircraft noise immissions around airfields and airports. This software is still used in Switzerland. It started with Basic on a TEK4052, moved to PCs and Pascal, Fortran 4 on mainframes and finally to C under GNU/Linux on PCs again where it does parallel processing using OpenMPI on my homebrew cluster. The way I maintained all that up to now is a great mess to say the least. Git might change that ...

IMG_0335

**Comment by djhoese** [comment]
I think all past concerns from this PR have been addressed. Let's get this merged and continue on reworking pycoast.

**Comment by howff** [comment]
Thanks very much David and Ernst for taking on extra work for this PR

**Pull request merged**
